src/payload.rs

```diff
@@ -0,0 +1,516 @@
+            PossiblyMissed::Entry(LogEntry::event(EventId::new(2).unwrap())),
+            PossiblyMissed::Entry(
+                LogEntry::clock(LogicalClock {
+                    id: probe_id,
+                    ticks: ProbeTicks(1),
+                    epoch: ProbeEpoch(0),
+                })
+                .0,
+            ),
+            PossiblyMissed::Entry(
+                LogEntry::clock(LogicalClock {
+                    id: probe_id,
+                    ticks: ProbeTicks(1),
+                    epoch: ProbeEpoch(0),
+                })
+                .1,
+            ),
+        ];
+
+        let mut payload = [0u8; 128];
+        let (num_written, num_read, did_clocks_overflow) = write_report_payload(
+            log.iter().copied(),
+            &mut FrontierClocksFSV(&mut frontier_clocks),
+            &mut PayloadOutputByteSlice::new(&mut payload),
+        );
+        assert_eq!(num_written, 6);
+        assert_eq!(num_read, 8);
+        assert!(!did_clocks_overflow);
+        let expected_payload = [
+            LogEntry::clock(LogicalClock {
+                id: probe_id,
+                ticks: ProbeTicks(0),
+                epoch: ProbeEpoch(0),
+            })
+            .0,
+            LogEntry::clock(LogicalClock {
+                id: probe_id,
+                ticks: ProbeTicks(0),
+                epoch: ProbeEpoch(0),
+            })
+            .1,
+            LogEntry::event_with_payload(EventId::EVENT_LOG_ITEMS_MISSED, 4).0,
+            LogEntry::event_with_payload(EventId::EVENT_LOG_ITEMS_MISSED, 4).1,
+            LogEntry::event(EventId::new(1).unwrap()),
+            LogEntry::event(EventId::new(2).unwrap()),
+            LogEntry::clock(LogicalClock {
+                id: probe_id,
+                ticks: ProbeTicks(1),
+                epoch: ProbeEpoch(0),
+            })
+            .0,
+            LogEntry::clock(LogicalClock {
+                id: probe_id,
+                ticks: ProbeTicks(1),
+                epoch: ProbeEpoch(0),
+            })
+            .1,
+        ];
+        let mut expected_payload_arr = [0u8; 24 * size_of::<LogEntry>()];
+        let expected_payload_len = log_to_bytes(&expected_payload, &mut expected_payload_arr);
+        let expected_payload_raw: &[u8] =
+            unsafe { core::mem::transmute(&expected_payload_arr[..expected_payload_len]) };
+        // Payload correct
+        let num_frontier_clocks = 1;
+        assert_eq!(
+            &payload[..(num_written + num_frontier_clocks * 2) * size_of::<LogEntry>()],
+            expected_payload_raw
+        );
+        // Recorded clocks correctly
+        assert_eq!(
+            frontier_clocks.as_slice(),
+            &[LogicalClock {
+                id: probe_id,
+                ticks: ProbeTicks(1),
+                epoch: ProbeEpoch(0),
+            }],
+        )
+    }
+
+    /// No space for internal event, don't write anything
+    #[test]
+    fn no_space_for_missed() {
```

Comment:
Could we also get a test for the other semi-error case, when there wasn't enough room in the frontier clock vec-like object to add a merged clock to it and subsequently a warning event was recorded in the log?